### PR TITLE
Fix treeview item icon

### PR DESCRIPTION
**Pull request by thegecko**
Signed-off-by: thegecko <rob.moran@arm.com>

#### What it does

The [label-provider](https://github.com/eclipse-theia/theia/blob/master/packages/core/src/browser/label-provider.ts#L369) class seems to iterate contributions for specified functionality until something is returned.

Both the [icon-theme-service](https://github.com/eclipse-theia/theia/blob/master/packages/core/src/browser/icon-theme-service.ts#L73) and [plugin-tree-view-node-label-provider](https://github.com/eclipse-theia/theia/blob/master/packages/plugin-ext/src/main/browser/view/plugin-tree-view-node-label-provider.ts#L42) contribution offer a `getIcon()` function.

Multiple contributions offering features seems to be mitigated by checking the return value is not `undefined`, however `icon-theme-service.getIcon()` returns an empty string instead and seems to get called first. This means all treeview icons are always empty.

This PR adds the additional check for an empty string which fixes the issue and the correct contribution is called, but I can't help feel this architecture is a messy approach and needs some further attention.

#### How to test

Load an extension with a treeview which has items with icons.

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/core/src/browser/label-provider.ts [hunk 1]
```diff
@@ -369,7 +369,7 @@ export class LabelProvider implements FrontendApplicationContribution {
     protected handleRequest(element: object, method: keyof Omit<LabelProviderContribution, 'canHandle' | 'onDidChange' | 'affects'>): string | undefined {
         for (const contribution of this.findContribution(element, method)) {
             const value = contribution[method]?.(element);
-            if (value !== undefined) {
+            if (value !== undefined && value !== '') {
```
Comment (colin-grant-work) [inline packages/core/src/browser/label-provider.ts:372]: Shall we just go to `!value` and exclude `null`, as well?

Comment (thegecko) [reply packages/core/src/browser/label-provider.ts]: Yup, done

**Review by colin-grant-work — APPROVED** [review]
Looks good to me; it is correct to treat an empty string as a non-return in these cases.

**Comment by vince-fugnitto** [comment]
@thegecko @colin-grant-work I think we might need to revert the changes for now, I noticed a regression (and possibly others) where the file-icon themes are not properly applied (ex: the theme is set to `none`):


_before_:

icons-before

_after_:

themes-after

**Comment by thegecko** [comment]
Recommend we investigate a proper fix rather than swapping one issue for another.

**Comment by thegecko** [comment]
I'll investigate, it looks like the file icon theme is being applied to all treeviews, not just the explorer.

**Comment by vince-fugnitto** [comment]
@thegecko what was the initial pull-request trying to achieve? I believe that `''` was a valid icon in terms that we did not want an icon to be displayed (perhaps we can use something like `'none'`. Anyone that called `this.labelProvider.getIcon()` would be returned an icon for the contribution that can best handle the element.

**Comment by colin-grant-work** [comment]
The problem (if it is one) is the entire system of [`ContributionProvider`s](https://github.com/eclipse-theia/theia/issues/10877) and unranked contributions. Anything we try to do here will be ad hoc, because there's no general way to order unordered contributions to make sure that we don't bail before getting the return we really want. Arguably the correct approach downstream is to inject providers you're interested in directly into the label provider and then check them in the order you want them checked before falling back to the default implementation, but that doesn't fix the fact that the default implementation implicitly relies on a specific ordering and undocumented sentinel values.

**Comment by vince-fugnitto** [comment]
@thegecko if that's the case then we should revert, tree-views should respect the file-icon theme, and in this case it was set to `none`. It should be the same behavior present in vscode as well.

**Comment by vince-fugnitto** [comment]
@thegecko what is the behavior you noticed in vscode when setting the theme to `none`?
I can't really test your extension, or take a look the sources.

**Comment by marcdumais-work** [comment]
@thegecko could you provide an example FOSS vscode extension that can be used to test this PR?

According the the project's [PR rules and expectations](https://github.com/eclipse-theia/theia/blob/master/doc/pull-requests.md#reverting), regressions introduced in a PR need to be addressed within 2 days or the PR can be reverted. Ideally we come to an agreement before tomorrow's release or risk delaying it.

Reading the comments above, it seems that a more generic fix could be ideal, and that in the meantime your product could be customized to achieve the result you want (we can assist if you want to explore that path).

**Comment by thegecko** [comment]
Without this patch, the default 'none' file tree theme doesn't show icons on any treeview:

Screenshot 2022-11-23 at 16 36 56

Switching the file tree theme then shows the treeview icons:

Screenshot 2022-11-23 at 16 37 11

While I agree this patch isn't the correct fix, I also believe we should only be affecting the file explorer treeview with the theme changes.

> I can't really test your extension

To recreate this, you can download the [device-manager vsix](https://open-vsx.org/api/arm/device-manager/1.2.5/file/Arm.device-manager-1.2.5.vsix) and put it in the `plugins` folder in theia and start the browser app.

You may need to modify the `Extensions > Device Manager > Manage All Devices` setting to show more USB devices.

> could you provide an example FOSS vscode extension that can be used to test this PR?

The extension above should be enough, it simply uses the `vscode.ThemeIcon` to set a treeview icon:

```typescript
    public getTreeItem(element: Node): vscode.TreeItem | Promise<vscode.TreeItem> {
        return {
            label: element.name,
            iconPath: new vscode.ThemeIcon('pass-filled')
        };
    }
```

> your product could be customized to achieve the result you want...

It would be quicker to implement a fix in Theia IMO

**Comment by colin-grant-work** [comment]
To supplement, here's the extension in VSCode with `none` theme set, with an icon still showing in this tree view:

image

**Comment by colin-grant-work** [comment]
The problem here seems to be that we don't respect icons on tree view elements at all:

https://github.com/eclipse-theia/theia/blob/909f4106e8c15c5c2c320401da4f48f8c6080734/packages/plugin-ext/src/main/browser/view/tree-view-widget.tsx#L432-L445

where `toNodeIcon` is the default implementation from the TreeWidget:

https://github.com/eclipse-theia/theia/blob/909f4106e8c15c5c2c320401da4f48f8c6080734/packages/core/src/browser/tree/tree-widget.tsx#L1411-L1413

**Comment by colin-grant-work** [comment]
@thegecko, I've proposed an alternative solution in #11914. What do you think of that?

**Pull request merged**